Harvester code cleanup

reaperrr opened the pull request:
Nothing major, but it's another small step towards cleaner code and might give some tiny performance improvements as well.

reaperrr committed:
Streamline Harvester.SetProcLines return check

reaperrr committed:
Cache INotifyHarvesterAction trait look-ups

reaperrr committed:
Clean up Harvester TickIdle a bit

Triggering UnblockRefinery each idle tick is bogus.

pchote commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -74,11 +74,6 @@ public override Activity Tick(Actor self)
 				self.QueueActivity(mobile.MoveTo(moveTo, 1));
 				self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
 
-				// TODO: The harvest-deliver-return sequence is a horrible mess of duplicated code and edge-cases
```
These notifications are correct in the context of what this code branch is trying to do: if no resources are found then the harvester should move to its assigned/assumed resource field instead of clogging up the base.  Having the unit drive there when it would normally use a carryall is unexpected.

This entire "get out of the way" feature, on the other hand, is of questionable use (there are other mechanisms designed to help unclog refineries) and is contributing to the overall mess of the harvester code.  If we're going to remove these notifications then we should go further and remove the whole code branch.  If this regresses gameplay in a meaningful way then we can reimplement it in a more maintainable way.

reaperrr replied:
`moveTo` is just some random free cell in the vincinity around the harvester, though, so `the harvester should move to its assigned/assumed resource field instead of clogging up the base` was awfully implicit if that was the original idea behind this code.

pchote replied:
The preferred value for `moveTo` is a random free cell in the vicinity of the last harvested resource cell.  It only falls back to the harvester location if there wasn't a harvested cell.

reaperrr replied:
>  If we're going to remove these notifications then we should go further and remove the whole code branch.

I'd prefer to remove only the "get out of the way" stuff, not the whole branch, because I have a follow-up that is using that branch.

pchote commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -118,20 +119,24 @@ public Harvester(Actor self, HarvesterInfo info)
 
 		void INotifyBuildComplete.BuildingComplete(Actor self)
 		{
+			notify = self.TraitsImplementing<INotifyHarvesterAction>().ToArray();
```
This duplication looks bogus.  Can we remover `BuildingComplete` completely from here?

reaperrr replied:
IIRC this was added to support deployable harvesters (YR slave miners), but added before stuff was moved to Mods.Common so it's not visible in the history.

Maybe @GraionDilach or @abcdefg30 knows/remembers more?

abcdefg30 replied:
Looks like it was added in #8674.

abcdefg30 replied:
The purpose was to make harvesters produced from factories automatically harvest. The question is if `BuildComplete` is still needed for that or if `Created` is enough.

reaperrr replied:
I remember now.
The 'move out of factory' overrides the FindResources queued by Created, that's what this was working around.

pchote replied:
Ok, then put `notify` *only* in Created, and the `QueueActivity` *only* in BuildingComplete.

reaperrr replied:
> Ok, then how about we put `notify` *only* in Created, and the `QueueActivity` *only* in BuildingComplete?

The former is fine, but the latter won't work because harvs spawned from refinery don't trigger BuildingComplete.

pchote replied:
Ok, in that case I have two suggestions:
* Try wrapping the QueueActivity from Created in a FrameEndTask to see if that makes it run late enough to not be overridden.
* If that fails, implement INotifyAddedToWorld, with a flag so that the code only runs the first time it is added.

Implementing BuildingComplete as a bandaid for the code run from Created not working all the time is ugly, and is something that we really should fix.

reaperrr replied:
> Try wrapping the QueueActivity from Created in a FrameEndTask to see if that makes it run late enough to not be overridden.

That seems to do the trick, will update accordingly.

reaperrr commented:
Updated.
Dropped the commit that removed target lines since it's not as straight-forward as I thought, and updated the lookup-caching commit so Harvester no longer implements INotifyBuildComplete.

pchote commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -279,7 +278,10 @@ void INotifyIdle.TickIdle(Actor self)
 			}
 
-			UnblockRefinery(self);
+			// We don't need to do this every idle tick, unblocking on first idle tick should be sufficient
```
I'm not sure I agree with this: the harvester may be boxed in and unable to move in the first tick, and then have a route open up later.

pchote reviewed (commented):
The other cleanups look good to me.

reaperrr commented:
Updated.

pchote commented:
Nothing to see here for normal players, so omitting a changelog entry.
